@CompileStatic causes StackOverflowError on Controller

### Issue description

When running `compileGroovy` on example project https://github.com/jdaugherty/grails-issue-stackoverflow-on-compile 

The following error occurs: 

        FAILURE: Build failed with an exception.
        
        * What went wrong:
          Execution failed for task ':compileGroovy'.
        > java.lang.StackOverflowError (no error message)

Here is the stack trace:
```
Caused by: java.lang.StackOverflowError
        at org.codehaus.groovy.ast.GenericsType.<init>(GenericsType.java:53)
        at org.codehaus.groovy.ast.GenericsType.<init>(GenericsType.java:57)
        at org.codehaus.groovy.ast.tools.GenericsUtils.correctToGenericsSpecRecurse(GenericsUtils.java:435)
        at org.codehaus.groovy.ast.tools.GenericsUtils.correctToGenericsSpecRecurse(GenericsUtils.java:371)
        at org.codehaus.groovy.ast.tools.GenericsUtils.extractSuperClassGenerics(GenericsUtils.java:561)
        at org.codehaus.groovy.ast.tools.GenericsUtils.parameterizeType(GenericsUtils.java:294)
        at org.codehaus.groovy.ast.tools.GeneralUtils.addAllInterfaces(GeneralUtils.java:482)
        at org.codehaus.groovy.ast.tools.GeneralUtils.addAllInterfaces(GeneralUtils.java:484)
        at org.codehaus.groovy.ast.tools.GeneralUtils.getInterfacesAndSuperInterfaces(GeneralUtils.java:495)
        at org.codehaus.groovy.ast.tools.WideningCategories.lowestUpperBound(WideningCategories.java:422)
        at org.codehaus.groovy.ast.tools.WideningCategories.lowestUpperBound(WideningCategories.java:210)
        at org.codehaus.groovy.ast.tools.WideningCategories.parameterizeLowestUpperBound(WideningCategories.java:276)
        at org.codehaus.groovy.ast.tools.WideningCategories.lowestUpperBound(WideningCategories.java:232)
        at org.codehaus.groovy.ast.tools.WideningCategories.parameterizeLowestUpperBound(WideningCategories.java:276)
        at org.codehaus.groovy.ast.tools.WideningCategories.lowestUpperBound(WideningCategories.java:232)
        at org.codehaus.groovy.ast.tools.WideningCategories.parameterizeLowestUpperBound(WideningCategories.java:276)
        at org.codehaus.groovy.ast.tools.WideningCategories.lowestUpperBound(WideningCategories.java:232)
        at org.codehaus.groovy.ast.tools.WideningCategories.parameterizeLowestUpperBound(WideningCategories.java:276)
        at org.codehaus.groovy.ast.tools.WideningCategories.lowestUpperBound(WideningCategories.java:232)
        at org.codehaus.groovy.ast.tools.WideningCategories.parameterizeLowestUpperBound(WideningCategories.java:276)
        at org.codehaus.groovy.ast.tools.WideningCategories.lowestUpperBound(WideningCategories.java:232)
        ...
```

--- a/dependencies.gradle
+++ b/dependencies.gradle
@@ -38,7 +38,7 @@ ext {
             'jquery.version'                : '3.7.1',
             'objenesis.version'             : '3.4',
             'gradle-spock.version'          : '2.3-groovy-3.0',
-            'spring-boot.version'           : '3.5.6',
+            'spring-boot.version'           : '3.5.7',
     ]
 
     // Note: the name of the dependency must be the prefix of the property name so properties in the pom are resolved correctly
@@ -73,7 +73,7 @@ ext {
             'bootstrap.version'           : '5.3.7',
             'commons-codec.version'       : '1.18.0',
             'geb-spock.version'           : '8.0.0',
-            'groovy.version'              : '4.0.28',
+            'groovy.version'              : '4.0.29',
             'h2.version'                  : '2.3.232',
             'jackson.version'             : '2.19.1',
             'jquery.version'              : '3.7.1',

